common/include/pcl/pcl_types.h

```diff
@@ -0,0 +1,136 @@
+   * \tparam T Type of the object stored inside the shared_ptr
+   */
+  template <typename T>
+  using shared_ptr = boost::shared_ptr<T>;
+
+  using uint8_t [[deprecated("use std::uint8_t instead of pcl::uint8_t")]] = std::uint8_t;
+  using int8_t [[deprecated("use std::int8_t instead of pcl::int8_t")]] = std::int8_t;
+  using uint16_t [[deprecated("use std::uint16_t instead of pcl::uint16_t")]] = std::uint16_t;
+  using int16_t [[deprecated("use std::uint16_t instead of pcl::int16_t")]] = std::int16_t;
+  using uint32_t [[deprecated("use std::uint32_t instead of pcl::uint32_t")]] = std::uint32_t;
+  using int32_t [[deprecated("use std::int32_t instead of pcl::int32_t")]] = std::int32_t;
+  using uint64_t [[deprecated("use std::uint64_t instead of pcl::uint64_t")]] = std::uint64_t;
+  using int64_t [[deprecated("use std::int64_t instead of pcl::int64_t")]] = std::int64_t;
+  using int_fast16_t [[deprecated("use std::int_fast16_t instead of pcl::int_fast16_t")]] = std::int_fast16_t;
+
+  namespace detail {
+    template <std::size_t Bits, bool Signed = true>
+    struct int_type { using type = void; };
+    template <std::size_t Bits, bool Signed = true>
+    using int_type_t = typename int_type<Bits, Signed>::type;
+
+    template <>
+    struct int_type<8, true> { using type = std::int8_t; };
+    template <>
+    struct int_type<8, false> { using type = std::uint8_t; };
+    template <>
+    struct int_type<16, true> { using type = std::int16_t; };
+    template <>
+    struct int_type<16, false> { using type = std::uint16_t; };
+    template <>
+    struct int_type<32, true> { using type = std::int32_t; };
+    template <>
+    struct int_type<32, false> { using type = std::uint32_t; };
+    template <>
+    struct int_type<64, true> { using type = std::int64_t; };
+    template <>
+    struct int_type<64, false> { using type = std::uint64_t; };
+  }  // namespace detail
+
+// temporary macros for customization. Only use for PCL < 1.12
+// Aim is to remove macros and instead allow multiple index types to coexist together
+#ifndef PCL_INDEX_SIZE
+#if PCL_MINOR_VERSION <= 11
+#define PCL_INDEX_SIZE sizeof(int)
+#else
+#define PCL_INDEX_SIZE 32
+#endif  // PCL_MINOR_VERSION
+#endif  // PCL_INDEX_SIZE
+
+#ifndef PCL_INDEX_SIGNED
+#define PCL_INDEX_SIGNED true
+#endif
+
+  /**
+   * \brief number of bits in PCL's index type
+   *
+   * For PCL 1.11, please use PCL_INDEX_SIZE to choose a size best suited for your needs.
+   * PCL 1.12 will come with default 32, along with client code compile time choice
+   *
+   * PCL 1.11 has a default size = sizeof(int)
+   */
+  constexpr short int index_type_size = PCL_INDEX_SIZE;
+
+  /**
+   * \brief signed/unsigned nature of PCL's index type
+   * For PCL 1.11, please use PCL_INDEX_SIGNED to choose a type best suited for your needs.
+   * PCL 1.12 will come with default signed, along with client code compile time choice
+   * Default: signed
+   */
+  constexpr bool index_type_signed = PCL_INDEX_SIZE;
```

Comment:
For some reason, I'm not very happy for exposing these two.

Reply:
I'm thinking of a 2 step process: 
1. we'll introduce `pcl::index_t` and `std::vector<pcl::index_t>` which is configurable (in c++14). This gives PCL compile time configuration
2. we'll remove the flexibility since in c++17 `pcl::index_t<>` can be written as `pcl::index_t` allowing `pcl::index_t` and `pcl::index_t<64, true>` to coexist. This allows client library compile time choice. Then it makes sense to query what `pcl::index_t` has as defaults. I don't like macros hence `constexpr` (which plays well with `if constexpr`)

Even if we do it on one swoop, we need the ability to query the defaults

Reply:
>  Then it makes sense to query what `pcl::index_t` has as defaults.

But allowing a query or exposing convenience symbols are different things. You can query everything you need with `sizeof` and `std::is_signed`. I'm not saying its a blocker for me, but I'm just not convinced yet that we need/should expose these symbols.